airflow/migrations/versions/c306b5b5ae4a_switch_xcom_table_to_use_run_id.py

```diff
@@ -0,0 +1,140 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Switch XCom table to use ``run_id``.
+
+Revision ID: c306b5b5ae4a
+Revises: e655c0453f75
+Create Date: 2022-01-19 03:20:35.329037
+"""
+from typing import Sequence
+
+from alembic import op
+from sqlalchemy import Column, Integer, LargeBinary, MetaData, Table, select
+
+from airflow.migrations.db_types import TIMESTAMP, StringID
+
+# Revision identifiers, used by Alembic.
+revision = "c306b5b5ae4a"
+down_revision = "e655c0453f75"
+branch_labels = None
+depends_on = None
+
+
+metadata = MetaData()
+
+dagrun = Table(
+    "dag_run",
+    metadata,
+    Column("id", Integer, primary_key=True),
+    Column("dag_id", StringID(), nullable=False),
+    Column("run_id", StringID(), nullable=False),
+    Column("execution_date", TIMESTAMP, nullable=False),
+)
+
+
+def _get_new_xcom_columns() -> Sequence[Column]:
+    return [
+        Column("dagrun_id", Integer(), nullable=False),
+        Column("task_id", StringID(), nullable=False),
+        Column("key", StringID(length=512), nullable=False),
+        Column("value", LargeBinary),
+        Column("timestamp", TIMESTAMP, nullable=False),
+        Column("dag_id", StringID(), nullable=False),
+        Column("run_id", StringID(), nullable=False),
+    ]
+
+
+def _get_old_xcom_columns() -> Sequence[Column]:
+    return [
+        Column("key", StringID(length=512), nullable=False),
+        Column("value", LargeBinary),
+        Column("timestamp", TIMESTAMP, nullable=False),
+        Column("task_id", StringID(), nullable=False),
+        Column("dag_id", StringID(), nullable=False),
+        Column("execution_date", StringID(), nullable=False),
+    ]
+
+
+def upgrade():
+    """Switch XCom table to use run_id.
+
+    For performance reasons, this is done by creating a new table with needed
+    data pre-populated, adding back constraints we need, and renaming it to
+    replace the existing XCom table.
+    """
+    op.create_table("__airflow_tmp_xcom", *_get_new_xcom_columns())
+
+    xcom = Table("xcom", metadata, *_get_old_xcom_columns())
+    query = select(
+        [
+            dagrun.c.id,
+            xcom.c.task_id,
+            xcom.c.key,
+            xcom.c.value,
+            xcom.c.timestamp,
+            xcom.c.dag_id,
+            dagrun.c.run_id,
+        ],
+    ).select_from(
+        xcom.join(
+            dagrun,
+            xcom.c.dag_id == dagrun.c.dag_id,
+            xcom.c.execution_date == dagrun.c.execution_date,
```

Comment:
Ooops! This needs `and_()` -- only the first condition was the onclause, the second is being passed to `isouter`

Reply:
And this fixed the speed! It was seconds after making this change. 

I'll try bigger scale (in the a few millions range)

Reply:
Wooho!

Reply:
BTW. I am always surprised how ORM constructs to build queries make things unintuitive. If only not the various backends support writing a plain 'SELECT FROM` query would have been so much more straightforward.

Reply:
(This is especially true when you do not care about the "object" property of the data).